New Data Source: aws_elasticsearch_domain

marcboudreau opened the pull request:
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" comments, they generate extra noise for pull request followers and do not help prioritize the request

Fixes #7899

This PR adds an AWS ElasticsearchDomain data source that can be used to retrieve any of the attributes returned by the **describe-elasticsearch-domain** API call or tags returned by the **list-tags** API call, by simply providing the *domain_name*.

I've successfully ran the acceptance test (targeting my new test only).  However, I was unable to add a check for the tags.

I'm also including some missing punctuation in the AWS ElasticsearchDomain **resource** documentation page that I noticed.

ValFadeev commented:
Hi,
First of all thanks for your work on implementing this data source. I found myself needing it to read an endpoint of an existing domain. 
First I had to fix a dependency, as `normalizeJsonString` now lives in `"github.com/hashicorp/terraform/helper/structure"`
```
$ make build
==> Checking that code complies with gofmt requirements...
go install
# github.com/terraform-providers/terraform-provider-aws/aws
aws/data_source_aws_elasticsearch_domain.go:158:20: undefined: normalizeJsonString
gmake: *** [GNUmakefile:8: build] Error 2
```
 It then turned out that for ES domains created in a [VPC](https://aws.amazon.com/blogs/aws/amazon-elasticsearch-service-now-supports-vpc/) `endpoint` field is not populated

```

data "aws_elasticsearch_domain" "test" {
  domain_name = "existing-vpc-domain"
}

output "endpoint" {
  value = "${data.aws_elasticsearch_domain.test.endpoint}"
}
```
This is due to the fact that for domains deployed into a VPC the API returns a map instead of a plain field, where the endpoint is an entry with `vpc` key:
```
$ aws es describe-elasticsearch-domain --domain-name existing-vpc-domain --query 'DomainStatus.Endpoints'
{
    "vpc": "vpc-xxxx.es.amazonaws.com"
}
```
I have made a few fixes [here](https://github.com/ValFadeev/terraform-provider-aws/commit/3b3c6c1773b12287588c377334bba5998fc16a79) and eventually got the expected output
```

data "aws_elasticsearch_domain" "test" {
  domain_name = "existing-vpc-domain"
}

output "endpoint" {
  value = "${element(values(data.aws_elasticsearch_domain.test.endpoints[0]), 0)}"
}
```

```
Outputs:

endpoint = vpc-xxxx.es.amazonaws.com
```
Happy to submit a PR to the feature branch if the changes look ok.

withnale commented:
Is this dead? It's been a while since this PR was introduced

marcboudreau commented:
Hi, sorry about this.  The issues that @ValFadeev pointed out were news to me.  I'm not sure if AWS introduced new options since I originally created the PR.

I'd need to spend time to study them before proceeding.

tchiak commented:
What's the status on this?

marcboudreau commented:
I've been taking a fresh look at this PR this week.  I'm updating the attributes of the datasource to match those of the resource.

marcboudreau commented:
I've updated this PR so that the **aws_elasticsearch_domain** data source now offers the same attributes (where it makes sense) as its resource counter-part.

I've run the acceptance tests by targeting the specific tests that I've introduced against my own personal AWS account.  I'm relunctant to run the entire suite, not really knowing what kind of cost 💰  that could incur.  To do so, I ran `TESTARGS="-run TestAccAWSDataElasticsearchDomain_advanced" make testacc` from the repository's root directory.

ryndaniels commented on website/docs/d/elasticsearch_domain.html.markdown:
```diff
@@ -0,0 +1,77 @@
+---
```
In addition to this new file, you'll also want to add a link to `aws.erb` in the `website` dir - those should be kept in alphabetical order but should be pretty straightforward to figure out 🙂

marcboudreau commented:
I've incorporated the feedback from the comments, and had to make some additional adjustments to the schema definition of the datasource (adding **zone_awareness_config** to the **cluster_options**).

Here are the outputs from the 2 acceptance test functions that I'm introducing:
*Basic test*
```
$ make testacc TEST=./aws TESTARGS='-run=TestAccAWSDataElasticsearchDomain_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -parallel 20 -run=TestAccAWSDataElasticsearchDomain_basic -timeout 120m
=== RUN   TestAccAWSDataElasticsearchDomain_basic
=== PAUSE TestAccAWSDataElasticsearchDomain_basic
=== CONT  TestAccAWSDataElasticsearchDomain_basic
--- PASS: TestAccAWSDataElasticsearchDomain_basic (802.63s)
PASS
ok      github.com/terraform-providers/terraform-provider-aws/aws       802.680s
```
*Advanced test*
```
$ make testacc TEST=./aws TESTARGS='-run=TestAccAWSDataElasticsearchDomain_advanced'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -parallel 20 -run=TestAccAWSDataElasticsearchDomain_advanced -timeout 120m
=== RUN   TestAccAWSDataElasticsearchDomain_advanced
=== PAUSE TestAccAWSDataElasticsearchDomain_advanced
=== CONT  TestAccAWSDataElasticsearchDomain_advanced
--- PASS: TestAccAWSDataElasticsearchDomain_advanced (1075.80s)
PASS
ok      github.com/terraform-providers/terraform-provider-aws/aws       1075.850s
```

ryndaniels commented on aws/data_source_aws_elasticsearch_domain.go:
```diff
@@ -0,0 +1,366 @@
+							Computed: true,
+						},
+						"zone_awareness_config": {
+							Type:     schema.TypeSet,
```
```suggestion
							Type:     schema.TypeList,
```

This should probably be `TypeList` to be consistent with the data type in the resource (see [here](https://github.com/terraform-providers/terraform-provider-aws/blob/master/aws/resource_aws_elasticsearch_domain.go#L194) for reference) 🙂

ryndaniels commented on website/aws.erb:
```diff
@@ -29,9 +29,2169 @@
                 <li>
                     <a href="#">Provider Data Sources</a>
                     <ul class="nav">
+                        <li>
```
It looks like this needs a rebase! 🙂

ryndaniels commented:
Thanks again @marcboudreau! I went ahead and made some changes to the `aws.erb` myself to get it in line with the recent changes on master, since I know there was a lot to deal with there. Good catch on adding the `zone_awareness_config` to the docs as well. 🙂 

Acceptance testing output:
```
--- PASS: TestAccAWSDataElasticsearchDomain_basic (746.56s)
--- PASS: TestAccAWSDataElasticsearchDomain_advanced (1545.41s)
```
